crates/pipeline-manager/src/runner/local_runner.rs

```diff
@@ -127,33 +117,162 @@ pub async fn fetch_binary_ref(
+    /// from stdout/stderr. New followers are caught up and existing followers receive
+    /// the new stdout/stderr lines as they come in.
+    ///
+    /// Returns a sender to invoke termination of the thread and the corresponding join handle.
+    fn setup_log_operational(
+        pipeline_id: PipelineId,
+        stdout: ChildStdout,
+        stderr: ChildStderr,
+        mut log_follow_request_receiver: Receiver<Sender<LogMessage>>,
+    ) -> (Sender<()>, JoinHandle<Receiver<Sender<LogMessage>>>) {
+        let (terminate_sender, mut terminate_receiver) = mpsc::channel::<()>(10);
+        let join_handle = spawn(async move {
+            // Buffered line readers such that we can already segment it into lines,
+            // such that stdout and stderr can be interleaved without making it unreadable
+            let mut lines_stdout = BufReader::new(stdout).lines();
+            let mut lines_stderr = BufReader::new(stderr).lines();
+
+            // Buffer with the latest lines
+            let mut logs = LogsBuffer::new(
+                Self::LOGS_BUFFER_LIMIT_BYTE,
+                Self::LOGS_BUFFER_LIMIT_NUM_LINES,
+            );
+
+            // All parties interested in receiving the logs
+            let mut log_followers: Vec<Sender<LogMessage>> = Vec::new();
+
+            // Continues running until the logging is sent a termination message,
+            // or an I/O or send error is encountered
+            loop {
+                select! {
+                    // Terminate
+                    _ = terminate_receiver.recv() => {
+                        Self::end_log_of_followers(&mut log_followers, LogMessage::End("pipeline is being shutdown".to_string())).await;
+                        debug!("Terminating logging by request");
+                        break;
+                    }
+                    // New follower
+                    follower = log_follow_request_receiver.recv() => {
+                        if let Some(follower) = follower {
+                            Self::catch_up_and_add_follower(&mut logs, &mut log_followers, follower).await;
+                        } else {
+                            // The automata cannot recover from this as this is the only method to receive log followers from the runner.
+                            // Operational logging should not be active when the pipeline is shutdown, and only if the pipeline is
+                            // shutdown should it be deleted (which would cause this). As such, this is an error.
+                            error!("Log request channel closed during operational logging. Logs for this pipeline will no longer work.");
+                            break;
+                        }
+                    }
+                    // New stdout line
+                    line = lines_stdout.next_line() => {
+                        if let Ok(line) = line {
+                            if let Some(line) = line {
```

Comment:
when would this stuff ever be None? do we need an else branch

Reply:
The case is when stdout (or stderr) would return that there are no more lines: https://docs.rs/tokio/1.40.0/tokio/io/struct.Lines.html#method.poll_next_line -- Thanks! I'll add an error case for that as well :+1:

Reply:
Added variables `stdout_finished` and `stderr_finished` that track if either of them returned a `None`